File: packages-galerina/galerina-devtools-benchmarks/benchmarks/verified-native-operation/bench.go
package main

import (
	"encoding/json"
	"os"
	"sort"
	"time"
)

const (
	iterations = 1_000_000
	warmups    = 2
	samples    = 9
)

//go:noinline
func observedRead(value int32) int32 {
	return value
}

//go:noinline
func traverse(values []int32) (int32, int) {
	var last int32
	index := 0
	for index < iterations {
		last = observedRead(values[index])
		index++
	}
	return last, index
}

type benchmarkResult struct {
	Runtime             string  `json:"runtime"`
	Iterations          int     `json:"iterations"`
	Result              int32   `json:"result"`
	SamplesNS           []int64 `json:"samplesNs"`
	MedianNS            int64   `json:"medianNs"`
	OperationsPerSecond int64   `json:"operationsPerSecond"`
	Unit                string  `json:"unit"`
	AntiElision         string  `json:"antiElision"`
}

func main() {
	values := make([]int32, iterations)
	for index := range values {
		values[index] = int32(index)
	}

	for warmup := 0; warmup < warmups; warmup++ {
		last, observed := traverse(values)
		if last != 999_999 || observed != iterations {
			panic("REFUSED: warmup semantic mismatch")
		}
	}

	measured := make([]int64, 0, samples)
	for sample := 0; sample < samples; sample++ {
		started := time.Now()
		last, observed := traverse(values)
		elapsed := time.Since(started).Nanoseconds()
		if last != 999_999 || observed != iterations {
			panic("REFUSED: measured semantic mismatch")
		}
		if elapsed <= 0 {
			panic("REFUSED: timer resolution mismatch")
		}
		measured = append(measured, elapsed)
	}

	ordered := append([]int64(nil), measured...)
	sort.Slice(ordered, func(left, right int) bool { return ordered[left] < ordered[right] })
	median := ordered[len(ordered)/2]
	result := benchmarkResult{
		Runtime:             "go",
		Iterations:          iterations,
		Result:              999_999,
		SamplesNS:           measured,
		MedianNS:            median,
		OperationsPerSecond: (iterations * 1_000_000_000) / median,
		Unit:                "element-reads/s",
		AntiElision:         "noinline-observed-read-per-index-and-observed-last-value",
	}
	if err := json.NewEncoder(os.Stdout).Encode(result); err != nil {
		panic(err)
	}
}

File: packages-ts/galerina-devtools-benchmarks/benchmarks/verified-native-operation/bench.rs
use std::convert::TryFrom;
use std::hint::black_box;
use std::time::Instant;

const ITERATIONS: usize = 1_000_000;
const WARMUPS: usize = 2;
const SAMPLES: usize = 9;

fn traverse(values: &[i32]) -> (i32, usize) {
    let mut index = 0usize;
    let mut last = 0i32;
    while index < ITERATIONS {
        last = black_box(values[index]);
        index += 1;
    }
    (last, index)
}

fn main() {
    let values: Vec<i32> = (0..ITERATIONS).map(|value| value as i32).collect();
    for _ in 0..WARMUPS {
        let (last, observed) = traverse(black_box(&values));
        if last != 999_999 || observed != ITERATIONS {
            panic!("REFUSED: warmup semantic mismatch");
        }
    }

    let mut samples: Vec<u64> = Vec::with_capacity(SAMPLES);
    for _ in 0..SAMPLES {
        let started = Instant::now();
        let (last, observed) = traverse(black_box(&values));
        let elapsed_u128 = started.elapsed().as_nanos();
        if last != 999_999 || observed != ITERATIONS || elapsed_u128 == 0 {
            panic!("REFUSED: measured semantic mismatch");
        }
        let elapsed = u64::try_from(elapsed_u128).expect("elapsed time must fit u64");
        samples.push(elapsed);
    }
    let mut ordered = samples.clone();
    ordered.sort_unstable();
    let median = ordered[ordered.len() / 2];
    let operations_per_second = ((ITERATIONS as u128) * 1_000_000_000u128)
        / (median as u128);
    let sample_json = samples
        .iter()
        .map(u64::to_string)
        .collect::<Vec<_>>()
        .join(",");
    println!(
        "{{\"runtime\":\"rust\",\"iterations\":{},\"result\":999999,\"samplesNs\":[{}],\"medianNs\":{},\"operationsPerSecond\":{},\"unit\":\"element-reads/s\",\"antiElision\":\"std::hint::black_box-per-index\"}}",
        ITERATIONS,
        sample_json,
        median,
        operations_per_second,
    );
}
